File: trees/convertBSTToGreaterTree.java
// https://leetcode.com/problems/convert-bst-to-greater-tree/
package trees;

public class convertBSTToGreaterTree{
    public static void main(String[] args) {
        System.out.println("This is an Interactive problem please Test the Solution In Leetcode");
    
    }
}

class SolutionCBSTTGT {


    // Submitted by @Jiganesh


    // Runtime: 0 ms, faster than 100.00% of Java online submissions for Convert BST to Greater Tree.
    // Memory Usage: 42.6 MB, less than 72.73% of Java online submissions for Convert BST to Greater Tree.
    

    // TC: O(N)
    // SC: O(1)
    private int[] array = new int[1];
    
    public TreeNode convertBST(TreeNode root) {
        
        traverse(root);
        return root;
        
    }
    
    public void traverse(TreeNode root){
    if (root != null ){
        traverse(root.right);
        array[0]+=root.val;
        root.val= array[0];
        traverse(root.left);
        }
    }
}

File: trees/convertBSTToGreaterTree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:

    # Runtime: 82 ms, faster than 89.47% of Python3 online submissions for Convert BST to Greater Tree.
    # Memory Usage: 16.7 MB, less than 77.23% of Python3 online submissions for Convert BST to Greater Tree.
    
    def convertBST(self, root):
    
        
        def helper(root, array):
            
            if root:
                helper(root.right, array)
                root.val += array[0]
                array[0] = root.val
                helper(root.left, array)
                
            return root
                
        array =[0]
        return helper(root, array)
                